### PR TITLE
add checksum.ktest

### PR DESCRIPTION
**Pull request by jpsollie**
When more advanced checksum algorithms are implemented,
checks should be done to verify its correctness

Signed-off-by: jpsollie <janpieter.sollie@edpnet.be>

### CODE REVIEW

File: tests/bcachefs/checksum.ktest [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
+#tester script to verify checksum correctness of bcachefs filesystem.
+#this script requires a bcachefs-tools installation, and a sufficient (1GB) amount of RAM to create a loopback filesystem
+
+#create neccessary infrastructure in tmpfs
+mkdir /tmp/bcachefs_checksum_test
+mount -t tmpfs none /tmp/bcachefs_checksum_test
+mkdir /tmp/bcachefs_checksum_test/mounted
+dd if=/dev/zero of=/tmp/bcachefs_checksum_test/zerofile.bin bs=4096 count=65536
```
Comment (koverstreet) [inline tests/bcachefs/checksum.ktest:10]: Don't do this - just use config-scratch-devs to get a scratch device

File: tests/bcachefs/checksum.ktest [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+dd if=/dev/urandom of=/tmp/bcachefs_checksum_test/randfile.bin bs=4096 count=65536
+fallocate -l 1GiB /tmp/bcachefs_checksum_test/loopdir.bin
+uuid="$(bcachefs format /tmp/bcachefs_checksum_test/loopdir.bin | grep External\ UUID | rev | cut -d$'\t' -f1 | rev)"
```
Comment (koverstreet) [inline tests/bcachefs/checksum.ktest:13]: Cool, but this throws away all the other output from bcachefs format. Look at run_quiet as used by the other tests - this way we get the format output when running the tests in verbose mode.

File: tests/bcachefs/checksum.ktest [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+	umount /tmp/bcachefs_checksum_test/mounted
+	mount -t bcachefs -o loop /tmp/bcachefs_checksum_test/loopdir.bin /tmp/bcachefs_checksum_test/mounted
+	cmp /tmp/bcachefs_checksum_test/zerofile.bin /tmp/bcachefs_checksum_test/mounted/zerofile.bin
+	j=$?
+	error=$(($error+$j))
+	[ "$j" == "0" ] || echo "$i checksum zero file test failed"
+	cmp /tmp/bcachefs_checksum_test/randfile.bin /tmp/bcachefs_checksum_test/mounted/randfile.bin
+	j=$?
+	error=$(($error+$j))
+	[ "$j" == "0" ] || echo "$i checksum random file test failed"
```
Comment (koverstreet) [inline tests/bcachefs/checksum.ktest:46]: Tests run in errexit mode, so the test will just exit and report failure if cmp fails

File: tests/bcachefs/checksum.ktest [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+dd if=/dev/zero of=/tmp/bcachefs_checksum_test/zerofile.bin bs=4096 count=65536
+dd if=/dev/urandom of=/tmp/bcachefs_checksum_test/randfile.bin bs=4096 count=65536
+fallocate -l 1GiB /tmp/bcachefs_checksum_test/loopdir.bin
+uuid="$(bcachefs format /tmp/bcachefs_checksum_test/loopdir.bin | grep External\ UUID | rev | cut -d$'\t' -f1 | rev)"
+
+#mount the fake block device to a fake bcachefs folder
+mount -t bcachefs -o loop /tmp/bcachefs_checksum_test/loopdir.bin /tmp/bcachefs_checksum_test/mounted
+
+#get the list of available checksums
+list=$(cat /sys/fs/bcachefs/${uuid}/options/data_checksum)
```
Comment (koverstreet) [inline tests/bcachefs/checksum.ktest:19]: Please use a better variable name for this

**Review by koverstreet — CHANGES_REQUESTED** [review]
I'm not clear on what the test is trying to do - your comments imply that perhaps there's error injection going on (?) - or is it just a basic unit test?

If we're just checking that we're not barfing when using the different checksum types (a useful test to have!), all we need to do is write some data and then read it back. But, if you read it back naively you'll just read it from the kernel page cache, and not use the bcachefs read path at all - you'll need to use an O_DIRECT read with dd.

No clue what your cmp is actually testing for.

Also, this could go in single_device.ktest

File: tests/bcachefs/checksum.ktest (hunk 3 above)
Comment (jpsollie) [reply tests/bcachefs/checksum.ktest:46]: yes, and that's exactly what I was trying to do

**Comment by jpsollie** [comment]
This is indeed a unit test: The script dumps a random and zero-only file into a newly created bcachefs filesystem with a supported checksum, unmounts it, remounts it, and then sees if the file is still accessible.  
If the checksum algorithm is not working properly, the algorithm may be considered "broken" and bcachefs will respond with an IO error
I do not really see why you'd need the O_DIRECT flag here.

**Pull request closed without merging**